Adds Katello packer devel box

johnpmitsch opened the pull request:
Re-opening up https://github.com/theforeman/forklift/pull/941

A stable katello devel environment image that can be created with packer

johnpmitsch commented:
Currently figuring out some hostname issues and the workflow

jturel commented:
Just in case you haven't seen this, I ran into the following during installer run:
```
    katello-devel: TASK [foreman_installer : Run installer] ***************************************
    katello-devel: fatal: [default]: FAILED! => changed=true                                                                                                                 
    katello-devel:   cmd: |-                                                                                                                                                 
    katello-devel:     foreman-installer -v --scenario katello-devel  --disable-system-checks --katello-devel-scl-ruby=rh-ruby25 --katello-devel-admin-password changeme
    katello-devel:   delta: '0:00:08.823339'                                                                                                                                 
    katello-devel:   end: '2019-10-11 14:25:32.168386'                                                                                                                       
    katello-devel:   msg: non-zero return code                                                                                                                               
    katello-devel:   rc: 20                                                                                                                                                  
    katello-devel:   start: '2019-10-11 14:25:23.345047'                                                                                                                     
    katello-devel:   stderr: |-                                                                                                                                              
    katello-devel:     tput: unknown terminal "unknown"                                                                                                                      
    katello-devel:     tput: unknown terminal "unknown"                                                                                                                      
    katello-devel:     Reverse DNS failed. Looking up 127.0.0.1 gave localhost, expected to match katello-dev-stable.example.com
    katello-devel:   stderr_lines: <omitted>                                                                                                                                 
    katello-devel:   stdout: |-                                                                                                                                              
    katello-devel:     Your system does not meet configuration criteria                                                                                                      
    katello-devel:     [[32m INFO[0m [34m2019-10-11T14:25:32[0m [36mverbose[0m] Executing hooks in group pre_migrations
    katello-devel:     [[32m INFO[0m [34m2019-10-11T14:25:32[0m [36mverbose[0m] All hooks in group pre_migrations finished
    katello-devel:     [[32m INFO[0m [34m2019-10-11T14:25:32[0m [36mverbose[0m] Executing hooks in group boot
    katello-devel:     [[32m INFO[0m [34m2019-10-11T14:25:32[0m [36mverbose[0m] All hooks in group boot finished
    katello-devel:     [[32m INFO[0m [34m2019-10-11T14:25:32[0m [36mverbose[0m] Executing hooks in group init
    katello-devel:     [[32m INFO[0m [34m2019-10-11T14:25:32[0m [36mverbose[0m] All hooks in group init finished
    katello-devel:     [[32m INFO[0m [34m2019-10-11T14:25:32[0m [36mverbose[0m] Loading default values from puppet modules...
    katello-devel:     [[32m INFO[0m [34m2019-10-11T14:25:32[0m [36mverbose[0m] ... finished      
    katello-devel:     [[32m INFO[0m [34m2019-10-11T14:25:32[0m [36mverbose[0m] Executing hooks in group pre_values
    katello-devel:     [[32m INFO[0m [34m2019-10-11T14:25:32[0m [36mverbose[0m] All hooks in group pre_values finished
    katello-devel:     [[32m INFO[0m [34m2019-10-11T14:25:32[0m [36mverbose[0m] Running installer with args [["-v", "--scenario", "katello-devel", "--disable-system-checks",
 "--katello-devel-scl-ruby=rh-ruby25", "--katello-devel-admin-password", "changeme"]]                                                                                        
    katello-devel:     [[32m INFO[0m [34m2019-10-11T14:25:32[0m [36mverbose[0m] Installer finished in 1.610561176 seconds
    katello-devel:   stdout_lines: <omitted> 
```

johnpmitsch commented:
looks like this is now working :grin: I added instructions on how to use in the README in the packer directory. 

Looks like random hostname issues were DHCP related, I added some flags on the kickstart template network command and was also able to specify a hostname, these variables (aside from the hostname) were set on the ifconfig-eth0 config on a dev box of mine, so I figure they will work here
```
network --bootproto=dhcp --device=eth0 --onboot=on --hostname=katello-dev-stable.example.com
```
This box will need to have a stable hostname, which you can specify in the box definition (there is an example and instructions in the README)

I may update the naming to be `centos7-katello-devel-stable`, which is more in line with our naming convention.

This is the first step in having a stable dev box. I envision the next step to be automate building this every day and then publish to vagrant cloud only if it succeeds. Then, have a box definition in forklift that always pulls the image from vagrant cloud. So for developers, you can just `vagrant up centos7-katello-devel-stable` and have a stable dev box in 5ish minutes. 

There will be some things to figure out as far as supporting git remotes, nfs, and other dev-specific configuration, but the traditional workflow won't be broken so that can still be used until that is worked out.

To test this out, just follow the instructions in the README to build the box locally and use it in a box definition. Let me know if there are any questions :smiley:

jturel commented:
Works for me too. Thanks for adding docs on usage.

johnpmitsch commented:
Thanks for testing all, will wait on @ehelms or another maintainer to approve before merging

johnpmitsch commented:
I uploaded an image to [vagrant cloud](https://app.vagrantup.com/katello/boxes/katello-devel) which you can try out with this box definition:

```
centos7-katello-devel-stable:
  memory: 10000
  box_name: katello/katello-devel
  hostname: centos7-katello-devel-stable.example.com
```

jturel commented:
I'll try out that definition! I guess it can be added to 99-local example ?

johnpmitsch commented:
> I'll try out that definition! I guess it can be added to 99-local example ?

@jturel  great! You can add to `vagrant/boxes.d/99-local.yaml` (the copied over one, not the `.example`) or create a new file in that folder

jturel commented:
> > I'll try out that definition! I guess it can be added to 99-local example ?
> 
> @jturel great! You can add to `vagrant/boxes.d/99-local.yaml` (the copied over one, not the `.example`) or create a new file in that folder

Oops, I meant as part of this PR!

johnpmitsch commented:
> > > I'll try out that definition! I guess it can be added to 99-local example ?
> > 
> > 
> > @jturel great! You can add to `vagrant/boxes.d/99-local.yaml` (the copied over one, not the `.example`) or create a new file in that folder
> 
> Oops, I meant as part of this PR!

ah yes, I was thinking a different PR, but I could now that something is up on vagrant cloud. I can a box and some docs explaining what it is and how it works.

johnpmitsch commented:
@ehelms @jturel I've updated with a box definition and some documentation explaining what it is and how it works. The big difference is creating a new box requires `vagrant box update` so the underlying box image is updated, feel free to try out this workflow (in the development docs) to make sure it works.

I also set up a cron job on my workstation to publish the box to vagrant cloud every day. I plan to move this to Jenkins, but it is much easier to set up locally to ensure the whole workflow works.

ehelms commented on docs/development.md:
````diff
@@ -57,6 +57,24 @@ cd foreman
+vagrant up centos7-katello-devel-stable
+```
+
+Once you have spun up the box, to create a new `centos-katello-devel-stable` box using the latest image from vagrant cloud you will need to update the underlying box image. For example:
````
This sentence is a bit confusing to me. You are trying to explain how to get a newer stable vagrant box? I would include a note that it will remove the existing environment as a warning rather than "updating" the existing which I think users might perceive.

johnpmitsch replied:
updated, let me know if its not clear now!

ehelms commented on docs/development.md:
````diff
@@ -57,6 +57,24 @@ cd foreman
+Once you have spun up the box, to create a new `centos-katello-devel-stable` box using the latest image from vagrant cloud you will need to update the underlying box image. For example:
+```
+vagrant destroy centos7-katello-devel-stable
+vagrant box update centos7-katello-devel-stable
````
I can't recall exactly. Does the libvirt plugin handle when `box update` is run properly? Or does an additional command have to be run?

johnpmitsch replied:
that is a good point, I updated the docs with the info to remove the old box definitions from both vagrant and libvirt, hopefully things are more clear now

ehelms commented on packer/README.md:
```diff
@@ -0,0 +1,38 @@
+## PACKER
```
This can be standard wording case: Packer

ehelms commented on packer/README.md:
```diff
@@ -0,0 +1,38 @@
+
+For example: `packer build centos7-katello-devel-stable.json`
+
+You can create a package image using the above command and then add that box to vagrant:
```
Whats a package image?

ehelms commented on packer/README.md:
````diff
@@ -0,0 +1,38 @@
+```
+centos7-katello-devel-stable:
+  box_name: centos7-katello-devel-stable
+  hostname: centos7-katello-devel-stable.example.com
````
Specifying hostname is required given we auto-generate a hostname that would conflict?

ehelms replied:
Can users set a custom hostname when deploying the stable box?

johnpmitsch replied:
Unfortunately no, since this is a box that has a katello install on it already, the hostname has to be fixed so the Katello install is under a hostname. The packer template overwrites the hostname with a `fix_hostname.sh` script so it is `centos7-katello-devel-stable.example.com` before katello is installed. 

Katello is then installed, the box is created and pushed to vagrant cloud. Then, when a dev spins up the box locally, they will need to create it with this hostname so it matches what Katello was installed with. This is why it is in the box definition

This does bring some limitations with multiple boxes. You could have multiple box definitions all using this hostname, but it will make DNS hard, you can only have one DNS entry for `centos7-katello-devel-stable.example.com` so that will have to be kept updated to the box you want it to point to. I believe you couldn't fudge the hostname DNS entry either because the cert won't match.

ehelms commented on vagrant/boxes.d/99-base-dev.yaml:
```diff
@@ -0,0 +1,72 @@
+---
```
These are moving out of the example file? If someone copies that example file they will have a clash of named boxes.

johnpmitsch replied:
ah sorry, good catch, that was just something I was trying out, didn't meant to include in PR. I'm

ehelms commented on vagrant/boxes.d/99-base-dev.yaml:
```diff
@@ -0,0 +1,72 @@
+    variables:
+      ssh_forward_agent: true
+      foreman_devel_github_push_ssh: True
+      katello_devel_github_username: <REPLACE ME>
```
This is the reason we kept these devel definitions in an example file, because we dont want users committing changes to configuration options such as this.

ehelms commented on vagrant/boxes.d/99-local.yaml.example:
```diff
@@ -69,3 +68,7 @@ centos7-dynflow-devel:
+
+centos7-katello-devel-stable:
+  box_name: centos7-katello-devel-stable
+  hostname: centos7-katello-devel-stable.example.com
```
Should users be optionally setting:

```
    variables:
      katello_devel_github_username: <REPLACE ME>
```

The crux of my question is, how do users using the stable box configure their forks automatically?

johnpmitsch replied:
I haven't figured out this part yet, I have to see how that works in forklift and if it can be done after the box is spun up.

Since the whole install is happening automatically and not by a single user, any customization will have to be done after the box (with katello already installed) is spun up, most likely in a playbook.

ehelms replied:
Might be a good reason to extract some capabilities like this to a devel playbook or encouraging the use of shared folders to the vagrant box.

johnpmitsch commented:
@ehelms updated with some clearer docs around how to use the box, update it, and clean out old box versions.

I also removed the box definition file I accidentally committed

johnpmitsch commented:
There is a lot of info here and in the docs, but the workflow for a dev is only:

```
# copy over the box definition by copying the whole file or manually copying and pasting the definition
cp vagrant/boxes.d/99-local.yaml.example  vagrant/boxes.d/99-local.yaml 

# spin up the box
vagrant up centos7-katello-devel-stable
```
